opensearch/src/main/java/org/opensearch/sql/opensearch/storage/scan/BackgroundSearchScanner.java

```diff
@@ -0,0 +1,177 @@
+import org.opensearch.sql.opensearch.client.OpenSearchClient;
+import org.opensearch.sql.opensearch.request.OpenSearchRequest;
+import org.opensearch.sql.opensearch.response.OpenSearchResponse;
+
+/**
+ * Utility class for asynchronously scanning an index. This lets us send background requests to the
+ * index while we work on processing the previous batch.
+ *
+ * <h2>Lifecycle</h2>
+ *
+ * The typical usage pattern is:
+ *
+ * <pre>
+ *   1. Create scanner: new BackgroundSearchScanner(client)
+ *   2. Start initial scan: startScanning(request)
+ *   3. Fetch batches in a loop: fetchNextBatch(request, maxWindow)
+ *   4. Close scanner when done: close()
+ * </pre>
+ *
+ * <h2>Async vs Sync Behavior</h2>
+ *
+ * The scanner attempts to operate asynchronously when possible to improve performance:
+ *
+ * <ul>
+ *   <li>When async is available (client has thread pool access): - Next batch is pre-fetched while
+ *       current batch is being processed - Reduces latency between batches
+ *   <li>When async is not available (client lacks thread pool access): - Falls back to synchronous
+ *       fetching - Each batch is fetched only when needed
+ * </ul>
+ *
+ * <h2>Termination Conditions</h2>
+ *
+ * Scanning will stop when any of these conditions are met:
+ *
+ * <ul>
+ *   <li>An empty response is received (lastBatch = true)
+ *   <li>Response is an aggregation or count response (fetchOnce = true)
+ *   <li>Response size is less than maxResultWindow (fetchOnce = true)
+ * </ul>
+ *
+ * Note: This class should be explicitly closed when no longer needed to ensure proper resource
+ * cleanup.
+ */
+public class BackgroundSearchScanner {
+  private final OpenSearchClient client;
+  @Nullable private final Executor backgroundExecutor;
+  private CompletableFuture<OpenSearchResponse> nextBatchFuture = null;
+  private boolean stopIteration = false;
+
+  public BackgroundSearchScanner(OpenSearchClient client) {
+    this.client = client;
+    // We can only actually do the background operation if we have the ability to access the thread
+    // pool. Otherwise, fallback to synchronous fetch.
+    if (client.getNodeClient().isPresent()) {
+      this.backgroundExecutor =
+          client.getNodeClient().get().threadPool().executor(SQL_BACKGROUND_THREAD_POOL_NAME);
+    } else {
+      this.backgroundExecutor = null;
+    }
+  }
+
+  private boolean isAsync() {
+    return backgroundExecutor != null;
+  }
+
+  /**
+   * @return Whether the search scanner has fetched all batches
+   */
+  public boolean isScanDone() {
+    return stopIteration;
+  }
+
+  /**
+   * Initiates the scanning process. If async operations are available, this will trigger the first
+   * background fetch.
+   *
+   * @param request The OpenSearch request to execute
+   */
+  public void startScanning(OpenSearchRequest request) {
+    if (isAsync()) {
+      nextBatchFuture =
+          CompletableFuture.supplyAsync(() -> client.search(request), backgroundExecutor);
```

Comment:
I am surprised this is working without copying the thread context  like this: https://github.com/opensearch-project/sql/blob/main/opensearch/src/main/java/org/opensearch/sql/opensearch/executor/OpenSearchQueryManager.java#L39

Reply:
What am I missing? Why do we need context copy in other file. Can you print the thread context in this task and see if it has user credentials in case of FGAC?

PPLPermissionsIT can we add a join test here.

Reply:
This isn't just used for joins, every query goes through this interface. I realized while benchmarking that there were no queries that weren't hitting the pool. That the security ITs pass means either this works or we don't have security ITs.

I believe it works because we supply the cluster settings during the construction of the executor, so it's built-in to the thread context (as opposed to starting a fresh thread with no executor)